cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,664 @@
+			cmpOp = seccomp.CompareNotEqual
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "<") {
+			cmpOp = seccomp.CompareLess
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// Default to adding a precise match if possible. Otherwise
+	// let seccomp figure out the architecture specifics.
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// used to mock in tests
+var (
+	archUbuntuArchitecture       = arch.UbuntuArchitecture
+	archUbuntuKernelArchitecture = arch.UbuntuKernelArchitecture
+)
+
+// For architectures that support a compat architecture, when the
+// kernel and userspace match, add the compat arch, otherwise add
+// the kernel arch to support the kernel's arch (eg, 64bit kernels with
+// 32bit userspace).
+func addSecondaryArches(secFilter *seccomp.ScmpFilter) error {
+	// note that all architecture strings are in the dpkg
+	// architecture notation
+	var compatArch seccomp.ScmpArch
+
+	// common case: kernel and userspace have the same arch. We
+	// add a compat architecture for some architectures that
+	// support it, e.g. on amd64 kernel and userland, we add
+	// compat i386 syscalls.
+	if archUbuntuArchitecture() == archUbuntuKernelArchitecture() {
+		switch archUbuntuArchitecture() {
+		case "amd64":
+			compatArch = seccomp.ArchX86
+		case "arm64":
+			compatArch = seccomp.ArchARM
+		case "ppc64":
+			compatArch = seccomp.ArchPPC
+		}
+	} else {
+		// less common case, kernel and userspace
+		// different. This can happen when running e.g. a
+		// amd64 kernel with a i386 userland. However in this
+		// case snapd would also only request i386 snaps. So
+		// the use-case is even stranger (from a distro
+		// perspective), a i386 snap would have to ship some
+		// 64bit code that would have to detect at runtime if
+		// it can be used or not. But we are told this is not
+		// rare with certain classes of embedded devices.
+		compatArch = UbuntuArchToScmpArch(archUbuntuKernelArchitecture())
+	}
+
+	if compatArch != seccomp.ArchInvalid {
+		return secFilter.AddArch(compatArch)
+	}
```

Comment:
This would always trigger a warning on e.g. i386. The "0" value of seccomp.ScmpArch is "0", so this is really the equivalent of the C code `if (compatArch > 0)`